Bluetooth: controller: Initial support for set periodic adv data

cvinayak opened the pull request:
Add initial support for LE Set Periodic Advertising Data
command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Initial support for set periodic adv data

Add initial support for LE Set Periodic Advertising Data
command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Set Periodic Advertising Data

Test the implementation of LE Set Periodic Advertising Data
command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -161,8 +161,115 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
+
+	/* Tx Power */
+	if (ter_hdr->tx_pwr) {
+		*--ter_dptr = *--ter_dptr_prev;
```
It looks kinda odd to me, that you decrement the data pointer to get the last octet in the header. To me, it would make a lot more sense to something like 
```
ter_hdr[TX_POWER_OFFSET] = ter_hdr_prev[TX_POWER_OFFSET]
```

Unless I'm misreading this line?

Thalley replied:
Or even better, since the struct has the field, why not do

```
ter_hdr->tx_pwr = ter_hdr_prev->tx_pwr;
```

cvinayak replied:
The values are present based on bitmap in the 8-bit header that precedes the values. Hence, fixed offset value is not applicable here. Structure member de-referencing cannot be used.

Thalley replied:
> Structure member de-referencing cannot be used.

Why not?

cvinayak replied:
![image](https://user-images.githubusercontent.com/6350656/98185957-06432900-1f0e-11eb-81b5-c595653a4e99.png)

> If a flag bit is set to 1, the corresponding Extended Header field is present;
otherwise, the corresponding Extended Header field is not present.

bit fields in Extended Header Flags define the presence of rest of the values, hence a fixed offset will not work.

Thalley replied:
Ah, it took a while, but I think I understand now. 

But I think you might have an off-by-one error here: As far as I can tell, you assign `ter_dptr` as
```
ter_dptr = (uint8_t *)ter_hdr + sizeof(*ter_hdr);
```

but do not increment the `ter_dptr` pointer any time before this statement. When you do `*--ter_dptr`, don't you overwrite the extended header flags, rather than writing to the octet that should contain the TxPower? It looks like you are missing an octet between the extended header flags, and the ACAD when you copy the AD data in line 244.

cvinayak replied:
> But I think you might have an off-by-one error here: As far as I can tell, you assign `ter_dptr` as
> 
> ```
> ter_dptr = (uint8_t *)ter_hdr + sizeof(*ter_hdr);
> ```
> 

`ter_dtpr` points to the byte after `ter_hdr`, it is not off-by-one.

>When you do `*--ter_dptr`, don't you overwrite the extended header flags, rather than writing to the octet that should contain the TxPower?

No, ter_dptr has been incremented in line in the `TODO` at line 212 ;-)

Thalley replied:
Ah :) 

I think it might be worth either adding a note here, or commenting the block until the `TODO` in 212 have been done then, as that certainly isn't clear from the current state

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -161,8 +161,115 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
+	ter_pdu_prev = lll_adv_sync_data_peek(lll_sync);
+	ter_com_hdr_prev = (void *)&ter_pdu_prev->adv_ext_ind;
+	ter_hdr = (void *)ter_com_hdr_prev->ext_hdr_adi_adv_data;
+	ter_hdr_prev = *ter_hdr;
```
I'm not sure I understand this line: You dereference a pointer and assign that another pointer? 

Shouldn't it have been 
```
ter_hdr_prev = ter_hdr;
```
?

cvinayak replied:
dereference a struct pointer  and assign to a struct (by-value). ter_hdr_prev  is used as read-only and later ter_hdr points to the double buffer that will get modified bit fields.

Thalley replied:
But `ter_hdr_prev` is a pointer, not a struct. Should it have been `*ter_hdr_prev = *ter_hdr;` then?

cvinayak replied:
no it is not:
`struct pdu_adv_hdr *ter_hdr, ter_hdr_prev;`

Thalley replied:
Ah, yes, you are correct. Had misread the type, as they were both defined on the same line. I would perhaps suggest declaring two variables of different types on different lines to avoid confusion :) You can resolve this comment

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -161,8 +161,115 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
+
+	/* AuxPtr */
+	if (ter_hdr_prev.aux_ptr) {
+		ter_dptr_prev -= sizeof(struct pdu_adv_aux_ptr);
```
`ter_dptr_prev` isn't being used further in this function. What's the point of this line?

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -161,8 +161,126 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
+	if (ter_hdr_prev.aux_ptr) {
+		ter_dptr_prev -= sizeof(struct pdu_adv_aux_ptr);
+
+		ull_adv_aux_ptr_fill(&ter_dptr, lll_sync->adv->phy_s);
```
Doesn't `ter_dptr` currently point to the TX power? Aren't we missing a `ter_dptr -= sizeof(struct pdu_adv_aux_ptr)`?

cvinayak replied:
The decrement is done in the function ``ull_adv_aux_ptr_fill`

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -430,6 +438,26 @@ static void test_advx_main(void)
 	k_sleep(K_MSEC(400));
 
+	printk("Update periodic advertising data 1...");
+	err = ll_adv_sync_ad_data_set(handle, AD_OP, sizeof(per_adv_data1),
+				      (void *)per_adv_data1);
+	if (err) {
+		goto exit;
+	}
+	printk("success.\n");
+
+	k_sleep(K_MSEC(400));
+
+	printk("Update periodic advertising data 2...");
+	err = ll_adv_sync_ad_data_set(handle, AD_OP, sizeof(per_adv_data2),
+				      (void *)per_adv_data2);
```
Should we test if the adv data is actually changed, by examining the data on the receiver? I guess this only tests whether the functions return true. Or is it too early to actually test that?

cvinayak replied:
It is verified in the PA report generation PR. But that PR depends on this PR, to set the PA data.

cvinayak committed:
Bluetooth: controller: Fix Extended Advertising data population

Fix Extended Advertising Data population when there is no
Extended Header to be filled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:
```diff
@@ -92,15 +92,16 @@ uint8_t ull_adv_aux_hdr_set_clear(struct ll_adv_set *adv,
 /* helper function to calculate common ext adv payload header length */
 static inline uint8_t
-ull_adv_aux_hdr_len_get(struct pdu_adv_com_ext_adv *com_hdr, uint8_t *dptr)
+ull_adv_aux_hdr_len_get(struct pdu_adv_com_ext_adv *com_hdr, uint8_t **dptr)
```
Not sure I like this change: Generally I'm of the conviction that a `get` function should never change any data/state of the parameters. I would revert this and increase the pointer outside of the function.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -161,8 +161,127 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
+	/* Set the tertiary PDU len */
+	ter_pdu->len = ter_len + len;
+
+	/* Check AdvData overflow */
```
ter_pdu->len is only 8 bits and CONFIG_BT_CTLR_ADV_DATA_LEN_MAX can be up to 1650, so I do not think this overflow check actually checks for an overflow.

cvinayak replied:
Fix for this is depending on https://github.com/zephyrproject-rtos/zephyr/pull/29448, hence will wait for it be merged.

cvinayak replied:
@wopu-ot updated with a 16 bit auto variable to check overflow.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:
```diff
@@ -92,15 +92,16 @@ uint8_t ull_adv_aux_hdr_set_clear(struct ll_adv_set *adv,
 		    sizeof(struct pdu_adv_hdr))) {
 		len = offsetof(struct pdu_adv_com_ext_adv,
 			       ext_hdr_adi_adv_data);
+		*dptr = (uint8_t *)com_hdr + len;
```
Renamed function and added comment.

cvinayak committed:
Bluetooth: controller: Fix AD data size overflow check

pdu->len is only 8 bits and CONFIG_BT_CTLR_ADV_DATA_LEN_MAX
can be upto 1650 bytes. Fix the implementation to use 16
bit auto variable to check AD data length overflow.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
